CLI tests & documentation

emilymdubois opened the pull request:
Refs https://github.com/mapbox/sumo/issues/5

- [x] Add CLI tests
- [x] Add documentation

@rclark, I was curious if you know of a way to test for 👍 JSON formatting if `cli.flags.grouped` or `cli.flags.json` are true. As the CLI functions operate now, this is piped directly to process.stdout and some cursory investigating leads me to think mocking `process.stdout` is tricky.

emilymdubois committed:
Adds tests for CLI functions

rclark commented on lib/cli.js:
```diff
@@ -0,0 +1,76 @@
+  });
+
+  const results = cli.flags.grouped ?
+    sumo.createReadStream('records', search) :
```
You could use a sinon spy on `sumo.createReadStream` to check that `--grouped` option leads to looking at records vs messages.

emilymdubois replied:
👍 I have an assertion [here](https://github.com/mapbox/sumo/blob/54a46038b2cd429bf875418dc88b28f802714f55/test/cli.test.js#L110) that confirms waving the `--grouped` flag leads to records being returned. I'll rearrange the code here to make things a little clearer. Right now it's hard to that that assertion matches [this test case](https://github.com/mapbox/sumo/blob/54a46038b2cd429bf875418dc88b28f802714f55/test/cli.test.js#L75).

rclark commented on lib/cli.js:
```diff
@@ -0,0 +1,76 @@
+    objectMode: true,
+    transform: function(obj, enc, callback) {
+      if (cli.flags.grouped || cli.flags.json)
+        return callback(null, `${JSON.stringify(obj)}\n`);
```
Woof testing in here is tough. You'd almost have to break out the `function(obj, enc, callback)`, export it, and unit test it. You're right that trying to spy on what's heading to `process.stdout` is probably very difficult.

emilymdubois committed:
Updates tests, adds formatting tests

emilymdubois committed:
Tweak

emilymdubois committed:
Adds documentation

emilymdubois commented:
@rclark, added some formatting tests & CLI documentation to the readme. Everything look 👍 ?

rclark commented on readme.md:
```diff
@@ -86,3 +86,42 @@ sumo.search(search, (err, data) => {
+* SUMO_LOGIC_ACCESS_ID environment variable
+* SUMO_LOGIC_ACCESS_KEY environment variable
+* `-q or --query` flag
+* `-f or --from` flag
```
I believe this is optional and defaults to 15m ago? If not then there's a discrepancy with documentation of the flag above.

emilymdubois replied:
@rclark I don't actually see where this defaults to 15m. Mind pointing out where on the `master` branch? I can add this logic if it's missing.

rclark replied:
It might just be that I read https://github.com/mapbox/sumo/blob/master/bin/sumo.js#L17 which *says* there's a 15m default, and it isn't actually implemented. In fact https://github.com/mapbox/sumo/blob/master/bin/sumo.js#L43 suggests that it is not.

We should either remove the line from the `--help` doc, or implement 15m as a default if that makes sense.

rclark commented on test/cli.test.js:
```diff
@@ -0,0 +1,164 @@
+'use strict';
```
:+1: :+1: the restructuring you've done here is awesome and the tests look great. The one last thing I would ask is that your assertions all have messages -- `assert.equal(found, expected, message)`.

I find it really difficult to track down what has actually failed in a set of tests where assertions don't have customized messages.

emilymdubois committed:
Adds test messages

emilymdubois committed:
Adds 15m default for --from flag

emilymdubois commented:
@rclark, added a 15m default for the --from flag since I think it makes the CLI easier to use. Also added some messages so the test output is easier to wade through.

This is good to 🚢  and release on my end unless you can think of anything else you'd like to roll in.

rclark commented on bin/sumo.js:
```diff
@@ -39,63 +39,5 @@ const cli = meow({
   string: ['query', 'from', 'to', 'duration']
```
@emilymdubois I believe you can actually set defaults here like 

```
default: {
  from: '15m'
}
```

emilymdubois committed:
Uses meow's native defaults option

rclark commented:
If this is good to go, go for the merge @emilymdubois! Then we can close #5.
